hive/agents/ppo.py

```diff
@@ -0,0 +1,384 @@
+        preprocessed_update_info = {
+            "observation": update_info["observation"],
+            "action": update_info["action"],
+            "reward": update_info["reward"],
+            "done": update_info["done"],
+        }
+        if "agent_id" in update_info:
+            preprocessed_update_info["agent_id"] = int(update_info["agent_id"])
+
+        return preprocessed_update_info
+
+    def preprocess_update_batch(self, batch):
+        """Preprocess the batch sampled from the replay buffer.
+
+        Args:
+            batch: Batch sampled from the replay buffer for the current update.
+
+        Returns:
+                - Preprocessed batch.
+        """
+        for key in batch:
+            batch[key] = torch.tensor(batch[key], device=self._device)
+        return batch
+
+    @torch.no_grad()
+    def act(self, observation):
+        """Returns the action for the agent.
+
+        Args:
+            observation: The current observation.
+        """
+
+        observation = torch.tensor(
+            np.expand_dims(observation, axis=0), device=self._device
+        ).float()
+        action, logprob, _, value = self._actor_critic(observation)
+        action = action.cpu().detach().numpy()
+        action = action[0]
+
+        self._logprob = logprob.cpu().numpy()
+        self._value = value.cpu().numpy()
+        return action
+
+    def update(self, update_info):
+        """
+        Updates the PPO agent.
+
+        Args:
+            update_info: dictionary containing all the necessary information to
+                update the agent. Should contain a full transition, with keys for
+                "observation", "action", "reward", and "done".
+        """
+        if not self._training:
+            return
+        if self._replay_buffer.ready():
+            self._replay_buffer.compute_advantages(self._value)
+            for _ in range(self._num_epochs):
+                valid_ind_size = self._replay_buffer._find_valid_indices()
+                num_batches = int(np.ceil(valid_ind_size / self._batch_size))
+                for _ in range(num_batches):
+                    batch = self._replay_buffer.sample(batch_size=self._batch_size)
+                    batch = self.preprocess_update_batch(batch)
+                    self._optimizer.zero_grad()
+
+                    _, _logprob, entropy, values = self._actor_critic(
+                        batch["observation"], batch["action"]
+                    )
+                    logratios = _logprob - batch["logprob"]
+                    ratios = torch.exp(logratios)
+                    advantages = batch["advantages"]
+                    if self._normalize_advantages:
+                        advantages = (advantages - advantages.mean()) / (
+                            advantages.std() + 1e-8
+                        )
+                    # Actor loss
+                    loss1 = -advantages * ratios
+                    loss2 = -advantages * torch.clamp(
```

Comment:
Is it possible to give these more descriptive names?

Reply:
yes named them as clipped and unclipped policy losses